pyopae/pycontext.cpp

```diff
@@ -0,0 +1,58 @@
+// Copyright(c) 2018, Intel Corporation
+//
+// Redistribution  and  use  in source  and  binary  forms,  with  or  without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of  source code  must retain the  above copyright notice,
+//   this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+// * Neither the name  of Intel Corporation  nor the names of its contributors
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "pycontext.h"
+
+buffer_registry* buffer_registry::instance_ = nullptr;
+
+buffer_registry::buffer_registry() : buffers_() {}
+
+buffer_registry& buffer_registry::instance() {
+  if (buffer_registry::instance_ == nullptr) {
+    buffer_registry::instance_ = new buffer_registry();
+  }
+  return *buffer_registry::instance_;
+}
+
+void buffer_registry::register_handle(handle_t handle) {
+  buffers_[handle] = buffer_list_t();
```

Comment:
buffers_[handle] is the start of the buffer list? buffer_list_t() is to create a list?

Reply:
buffers_ is a map (associative array) keyed by a handle::ptr_t object.
This line is inserting an empty vector of buffer::ptr_t objects which is used later when registering a buffer::ptr_t object

Reply:
Is the map also a c++ data type? Or it is defined by pybind?

Reply:
map is part of the c++ stl (standard library).
You get it by including `<map>` and is found in the `std` namespace.
```C++
#include <map>

int main(){
  std::map<char*, int> str_to_ints;
  str_to_ints["one"] = 1;
}
```